jolt-core/src/r1cs/builder.rs

```diff
@@ -0,0 +1,1443 @@
+            .enumerate()
+            .map(|(input_index, input_lc)| {
+                let values = if let Some(range) = self.input_to_flat[input_index].clone() {
+                    &values[range]
+                } else {
+                    &[]
+                };
+                input_lc.evaluate(values)
+            })
+            .collect();
+        (self.compute)(&computed_inputs)
+    }
+}
+
+pub struct R1CSBuilder<F: JoltField, I: ConstraintInput> {
+    constraints: Vec<Constraint<I>>,
+    pub next_aux: usize,
+    aux_computations: Vec<AuxComputation<F, I>>,
+}
+
+impl<F: JoltField, I: ConstraintInput> Default for R1CSBuilder<F, I> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<F: JoltField, I: ConstraintInput> R1CSBuilder<F, I> {
+    pub fn new() -> Self {
+        Self {
+            constraints: vec![],
+            next_aux: 0,
+            aux_computations: vec![],
+        }
+    }
+
+    fn allocate_aux(
+        &mut self,
+        symbolic_inputs: Vec<LC<I>>,
+        compute: Box<AuxComputationFunction<F>>,
+    ) -> Variable<I> {
+        let new_aux = Variable::Auxiliary(self.next_aux);
+        self.next_aux += 1;
+
+        let computation = AuxComputation::new(new_aux, symbolic_inputs, compute);
+        self.aux_computations.push(computation);
+
+        new_aux
+    }
+
+    /// Index of variable within z.
+    pub fn witness_index(&self, var: impl Into<Variable<I>>) -> usize {
+        let var: Variable<I> = var.into();
+        match var {
+            Variable::Input(inner) => inner.into(),
+            Variable::Auxiliary(aux_index) => I::COUNT + aux_index,
+            Variable::Constant => I::COUNT + self.next_aux,
+        }
+    }
+
+    pub fn constrain_eq(&mut self, left: impl Into<LC<I>>, right: impl Into<LC<I>>) {
+        // left - right == 0
+        let left: LC<I> = left.into();
+        let right: LC<I> = right.into();
+
+        let a = left - right.clone();
+        let b = Variable::Constant.into();
+        let constraint = Constraint {
+            a,
+            b,
+            c: LC::zero(),
+        };
+        self.constraints.push(constraint);
+    }
+
+    pub fn constrain_eq_conditional(
+        &mut self,
+        condition: impl Into<LC<I>>,
+        left: impl Into<LC<I>>,
+        right: impl Into<LC<I>>,
+    ) {
+        // condition  * (left - right) == 0
+        let condition: LC<I> = condition.into();
+        let left: LC<I> = left.into();
+        let right: LC<I> = right.into();
+
+        let a = condition;
+        let b = left - right;
+        let c = LC::zero();
+        let constraint = Constraint { a, b, c };
+        self.constraints.push(constraint);
+    }
+
+    pub fn constrain_binary(&mut self, value: impl Into<LC<I>>) {
+        let one: LC<I> = Variable::Constant.into();
+        let value: LC<I> = value.into();
+        // value * (1 - value)
+        let constraint = Constraint {
+            a: value.clone(),
```

Comment:
I don't think we need this clone, because of the `into()` above? Here and in the other `constrain` functions

Reply:
Unfortunately we don't have ops by reference. Super cheap though so going to leave it.